### PR TITLE
Add pull request guide

### PR DESCRIPTION
**Pull request by intrepidmatt**
This article is an outline to be filled in later.

### CODE REVIEW

File: pull_request_guidelines.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+## Merging Pull Requests
+
+* When to merge (how many people need to approve)
+ * Everyone on the team should approve before merging.
+* Size of the team vs. size of the PR
+* Rebase before merging with `--no-ff`
+* Who performs the merge?
+ * The requester should merge.
+* Should the github “big green button” be used?
+ * Yes, this is the preferred method of merging.
```
Comment (hchood) [inline pull_request_guidelines.md]: fwiw backend team doesn't use the big green button.  we rebase and merge locally.  historically due to DT & Edward not liking the merge commits it generates.

why is this the preference?

File: pull_request_guidelines.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * You should provide bullet points of the changes you are making.
+ * It is an opportunity to offer up any explanation for why you want to make these changes.
+ * Github allows you to use markdown in here, so format it well.
+* When should screenshots be included?
+ * This is usually not necessary in creating a request.
+ * Maybe if the fix you are making is impossible to explain in words?
+* [Git process, feature branches](git_workflow_quick_reference.md)
+* Assigning people vs. posting in slack
+ * When you're on a team, or working with a another specific dev, it's best to directly assign them.
+ * When the PR is more general, post in slack to get someone else's eyes on it.
```
Comment (hchood) [inline pull_request_guidelines.md:33]: Before you post in slack or assign a reviewer, make sure you review the PR yourself to catch any obvious errors.

Also, if you have an automated style-guide checker (in ruby we have `rubocop`) run it first and fix any style errors before posting.

Comment (hchood) [inline pull_request_guidelines.md:33]: FWIW, backend team does "WIP" (work-in-progress) and "RFR" (ready-for-review) PRs.  Once we've started on a feature, we put up a "WIP" PR so that others can see what we're doing.  Once it's complete, we rename it with the prefix "RFR".

It's also nice to reference the Jira ticket number in the PR name or description, if there is one.

For myself, I like to create bulleted `TODO` lists in "WIP" PRs to help myself manage tasks.  Eg.

TODO
- [ ] Update wiki
- [ ] Test edge cases in service objects
- [ ] Extract `DateFormatter` service

File: pull_request_guidelines.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * Formatting corrections. i.e. spacing and parentheses, spelling errors
+ * Structural suggestions. i.e. order of methods, refactoring ideas
+ * Obvious mistakes. i.e. unused variables or misused APIs
+ * Or just ask clarifying questions if something is confusing.
+ * There are no dumb comments, but there is dumb code.
+* Correctness
+ * There may not be a correct way to do something yet, either because it's not defined in our style guide or multiple ways are acceptable. Go with your gut!
+* When is it necessary to run the code?
+ * It never hurts...
+ * It should be run and tested when there are heavy UI changes or if you are unsure of any particular change.
```
Comment (hchood) [inline pull_request_guidelines.md:48]: If there are tests, run the test suite locally.

Comment (hchood) [inline pull_request_guidelines.md:48]: Comment on the PR itself and not on the specific commits.  Exception might be if you're reviewing changes made in response to your comments, and they are hard to locate in the PR itself vs. in the specific commit.

File: pull_request_guidelines.md (hunk 2 above)
Comment (ryanmeador) [inline pull_request_guidelines.md:33]: How is this process significantly different from just pushing your WIP branch but not creating a PR for it?  I'd rather see the PR feature used for actual PRs; coworkers that want to keep tabs on what you're working on can easily look at the diff of your branch.

File: pull_request_guidelines.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+## Style
+
+* When should we be nit-picky about style?
+ * ALWAYS!
+ * It makes you better and it makes your code easier for others to read.
+
+## Responding to Feedback
+
+* When to make changes
+ * Is it best to get all the comments from someone before making changes? Or should you make push fixes for comments as they come in?
```
Comment (ryanmeador) [inline pull_request_guidelines.md]: I think it's fine to push fixes for comments as they come in, but it should probably be done on a case by case basis (I could see a contentious comment on a larger team requiring more discussion before it's fixed). I think it's worth noting that when you push a fix that changes a line that has a comment on it, github hides the comment (it says something to the effect of "Ryan commented on an outdated diff"). For this reason, I like to make a very general overview of all my diff comments as a comment on the PR itself, which never gets hidden.

File: pull_request_guidelines.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+* How to update the PR (force push)
+ * Pushing new commits will update the PR automatically.
+ * If you're rebasing, you will have to force push.
```
Comment (gdaniels) [inline pull_request_guidelines.md:71]: Let's make sure to add the requisite HUGE WARNING TO BE CAREFUL WHEN FORCE-PUSHING.

File: pull_request_guidelines.md (hunk 1 above)
Comment (ryanmeador) [inline pull_request_guidelines.md]: Is the issue here that you don't like the auto-generated commit message, or that you don't like the merge commit at all (in which case you'd probably also be opposed to `--no-ff`)?

File: pull_request_guidelines.md (hunk 2 above)
Comment (hchood) [inline pull_request_guidelines.md:33]: I find it much easier to (a) see what others are working on, and (b) outline for myself the steps that must be completed for a feature.  Would rather do that planning / work up front and expose it to others than wait till the end.

File: pull_request_guidelines.md (hunk 1 above)
Comment (hchood) [inline pull_request_guidelines.md]: Didn't like the merge commit.

Personally, I don't much care one way or another.

File: pull_request_guidelines.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+* Maintaining consistency
+
+## Pull Request Software
+
+* Reviewboard vs. Github
+
+## Creating Pull Requests
+
+* Pull request granularity (how big should my pull requests be?)
+ * Try to keep them under 500 lines.
```
Comment (weberbry) [inline pull_request_guidelines.md:16]: Possibly address small PRs. Somewhat tied into the next point, "single responsibility principle", small PRs are ok as well, assuming they address a single issue. 

Starting out I know I felt that I sometimes needed to bundle small unrelated commits into a single PR, which in my opinion is not ideal.

File: pull_request_guidelines.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+* Pull request “single responsibility principle”
+ * Pull requests should have one focus, for example: Adding a new model. Or updating an API.
+ * This will also help keep them short.
+* Pull request description
+ * You should provide bullet points of the changes you are making.
+ * It is an opportunity to offer up any explanation for why you want to make these changes.
+ * Github allows you to use markdown in here, so format it well.
```
Comment (weberbry) [inline pull_request_guidelines.md:25]: Include link to Jira ticket, if it makes sense for the project

File: pull_request_guidelines.md (hunk 1 above)
Comment (ryanmeador) [inline pull_request_guidelines.md]: I prefer the merge commit because it keeps it obvious what the chunk of work was.  IIRC, if you don't have the merge commit/keep the history linear, there's no way within Git to see what the PR contents was, you have to go digging in Github.

File: pull_request_guidelines.md (hunk 6 above)
Comment (ryanmeador) [inline pull_request_guidelines.md:16]: It's probably a bit out of scope for this document, but it might be worthwhile to write up ways of rewriting history to separate multiple responsibilities that got accidentally lumped together.

In the same vein, we might want to address the right way to file multiple PRs at once (do you have one based off the other and then rebase onto master before the merge, or do you just let the magic green button handle it and get some slightly weird-looking history?)

File: pull_request_guidelines.md (hunk 1 above)
Comment (intrepidmatt) [inline pull_request_guidelines.md]: FWIW I also prefer a merge commit for this reason – makes reading history easier, with references to the PRs themselves.

Comment (gdaniels) [inline pull_request_guidelines.md]: +1 merge commit

File: pull_request_guidelines.md (hunk 4 above)
Comment (weberbry) [inline pull_request_guidelines.md]: If a comment results in an offline discussion and NOT a code change, should an additional comment be added to reflect the offline discussion so its clear to a subsequent reader why there was no code changed?

File: pull_request_guidelines.md (hunk 7 above)
Comment (ButkiewiczP) [inline pull_request_guidelines.md:25]: +1 to include Jira ticket if applicable to the project

File: pull_request_guidelines.md (hunk 1 above)
Comment (hchood) [inline pull_request_guidelines.md]: Fair point.  We shall change our ways. :)

File: pull_request_guidelines.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+* How to update the PR (force push)
+ * Pushing new commits will update the PR automatically.
+ * If you're rebasing, you will have to force push.
+* When master moves forward, when should the feature branch be rebased?
+ * Always? Or just when conflicts are created?
+
+## Merging Pull Requests
+
+* When to merge (how many people need to approve)
+ * Everyone on the team should approve before merging.
```
Comment (loganwright) [inline pull_request_guidelines.md]: I'm not sure I agree w/ this one, it works well for teams of 2, but if there's something like a team of 5, do they really need 4 approvals?

I think maybe something less strict might be good

**Comment by loganwright** [comment]
One thing I don't see that might be useful.  If a PR is approved and the requester makes non-comment specific changes, they should request a secondary review, or withhold those changes for a new PR

File: pull_request_guidelines.md (hunk 4 above)
Comment (paulrolfe) [inline pull_request_guidelines.md]: I think the PR comments themselves should explain why no code was changed. So if there's an online discussion about it, some result comment should be added to the PR to clarify.

File: pull_request_guidelines.md (hunk 2 above)
Comment (paulrolfe) [inline pull_request_guidelines.md:33]: What's the verdict on this? Should @hchood's system be part of the PR flow?

Comment (ryanmeador) [inline pull_request_guidelines.md:33]: I'm mildly opposed.  I feel like it's an abuse of the PR feature in GitHub and better tracked someplace like Jira if it needs to be public or a notebook if it doesn't need to be public.

File: pull_request_guidelines.md (hunk 8 above)
Comment (brightredchilli) [inline pull_request_guidelines.md]: +1. We're not going to be able to enforce this one on large teams.

File: pull_request_guidelines.md (hunk 2 above)
Comment (hchood) [inline pull_request_guidelines.md:33]: I don't think I was clear in my original comment.  We don't create multiple PRs per feature -- we rename a given PR from "WIP" to "RFR" when it is complete.  We're not creating duplicate PRs that are never merged in.

File: pull_request_guidelines.md (hunk 3 above)
Comment (paulrolfe) [inline pull_request_guidelines.md:48]: I addressed both of these on other lines. :+1:

File: pull_request_guidelines.md (hunk 8 above)
Comment (paulrolfe) [inline pull_request_guidelines.md]: I addressed this in the next line.

File: pull_request_guidelines.md (hunk 6 above)
Comment (paulrolfe) [inline pull_request_guidelines.md:16]: @weberbry I think I addressed this below.

@ryanmeador Maybe that first part is better suited to the `Git Workflow` section of sherpa. The second part could fit here, although it might be addressed below when we say to always rebase before merging.

File: pull_request_guidelines.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+* When is it necessary to look at the whole codebase rather than just the diff?
+ * If changes touch or reference a piece of code you're unfamiliar with -- go check it out!
+ * When files are added, you may want to go verify the file structure is still organized correctly.
+
+## Style
+
+* When should we be nit-picky about style?
+ * ALWAYS!
+ * It makes you better and it makes your code easier for others to read.
+* **Note on tone**: Please don't be offended by lots of style corrections and other comments. Reviewers are not being mean, they are just making the you and the code better. :)
```
Comment (alanscarpa) [inline pull_request_guidelines.md]: Typo "making the you"

File: pull_request_guidelines.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+* When to make changes
+ * Make easy changes as they come in, and push a commit with a few changes at once.
+ * If there's something that needs discussion, don't change it right away because it will hide the existing comments in the PR if you edit that line.
+* How to update the PR (force push)
+ * Pushing new commits will update the PR automatically.
+ * If you're rebasing, you will have to force push.
+   * WARNING: Force pushes re-write history, so be very careful that you are pushing to the correct branch.
+* When master moves forward, when should the feature branch be rebased?
+ * Always? Or just when conflicts are created?
```
Comment (alanscarpa) [inline pull_request_guidelines.md]: Looks like this wasn't answered yet.  I rebase whenever master moves forward.

File: pull_request_guidelines.md (hunk 8 above)
Comment (Benuuu) [inline pull_request_guidelines.md]: I think this depends heavily on the size of the PR.  I don't need 2 extra sets of eyes on a one line fix especially if the project is well tested.

**Comment by Benuuu** [comment]
:shipit: :one:

File: pull_request_guidelines.md (hunk 8 above)
Comment (vinnyt) [inline pull_request_guidelines.md]: size and gravity of project should dictate this IMHO.   trying to shoot for half team +1 when > 4 is my personal favorite, but as the team grows it will slow things down.  If the team is that large the project is probably big enough and complex enough to warrant the extra eyes.  All of that being said the author of the pull should be honest and use their best judgment to get the right amount of feedback before merging.
